include/iterlib/FilterIterator.h

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include "iterlib/WrappedIterator.h"
+
+namespace iterlib {
+
+enum class FilterType {
+  INVALID = 0,
+  GE,
+  GT,
+  LE,
+  LT,
+  RANGE,
+  EQ,
+  NE,
+  EXISTS,
+  PREFIX,
+  INSET,
+  CONTAINS,
+};
+
+class FilterIteratorBase : public WrappedIterator {
+ public:
+  explicit FilterIteratorBase(Iterator* iter) : WrappedIterator(iter) {}
+
+  bool orderPreserving() const override { return true; }
+
+ protected:
+  bool doNext() override;
+
+  bool doSkipTo(id_t id) override;
+
+  virtual bool match(const Iterator* iter) = 0;
+};
+
+class FilterIterator : public FilterIteratorBase {
+ public:
+  explicit FilterIterator(Iterator* iter)
+      : FilterIteratorBase(iter), firstTime_(true) {}
+
+  void setFilter(const std::vector<std::string>& fields,
+                 const std::vector<dynamic>& values, FilterType filterType) {
+    if (fields.size() == 1) {
+      fields_ = fields[0];
+    } else {
+      auto fieldVec = variant::vector_dynamic_t();
+      for (const auto& v : fields) {
+        fieldVec.emplace_back(v);
+      }
+      fields_ = std::move(fieldVec);
+    }
+
+    // We support operators with:
+    //  1 field, 1 value
+    //  1 field, n values
+    // and some combination of the two
+    //
+    // TODO: Better detection of incorrect arity of field/values
+    // Eg: (filter (= (a  b) (1 2 3)))
+    auto valueVec = variant::vector_dynamic_t();
+    int i = 0;
+    for (const auto& value : values) {
+      // Use other type information for attributes
+      // that are ints
+      if ((i < fields.size()) &&
```

Comment:
They're doing different types of validations even though the code looks similar.

`setFilter` is trying to do compile time validation, without knowing the type of each field in `fields`. It needs to allow the following cases:

1. id() INSET (1, 2, 3, 4, 5)
2. (= (a b c)  (1 2 3))
3. But disallow (= (a b c) (1 2 3 4 5))

`match()` on the other hand is doing more runtime validation where types are known. If you get:

1. (= a (1 2 3)) # a is a string
2. (inset b (1 2 3)) # b is not a vector type

it is expected to throw exceptions.

The main TODO here is to do better validation in setFilter to catch more incorrect queries at compile time. More precisely, compile time == iterator tree construction time.